### PR TITLE
[Blog Post] Apache Beam for a content discovery platform

### PR DESCRIPTION
**Pull request by prodriguezdefino**
A blog post entry showcasing the usage of Apache Beam in a streaming pipeline to extract and store document content for a GenAI based content discovery platform. Full code available at https://github.com/prodriguezdefino/content-dicovery-platform-gcp

### CODE REVIEW

**Comment by prodriguezdefino** [comment]
heads up: @nams113 @aaltay

**Comment by aaltay** [comment]
Thank you @prodriguezdefino !

@liferoad - Could someone help with resolving the staging errors? If they can point out the issues, authors here could fix it.
@rszper - If you get a chance, could you review the text?

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+-->
+
+# DIY GenAI Content Discovery Platform with Apache Beam
+
+## Introduction
+
+Your digital assets, such as documents, PDFs, spreadsheets, and presentations, contain a wealth of valuable information which sometimes is hard to find or discover. This blog post discusses how to build a DIY starter architecture, based on near real time ingestion processing and large language models (LLMs) to extract meaningful information from your assets, making it available and discoverable through a simple natural language query.
+
+Building a near real time processing pipeline for content ingestion may seem like a very complex task, and it can be. To alliviate this task Apache Beam framework exposes a set of powerful constructs to remove the complexities of interacting with a variety of content sources and destinations, error handling, clarity, modularity, all while maintaining a good degree of resiliency and scalability with minimal effort. An Apache Beam streaming pipeline should be able to connect to the many components of a solution to quickly process document's content ingestion requests, making the information available in few seconds after the ingestion request was triggered.
```
Comment (aaltay) [inline website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:29]: @prodriguezdefino - Looks like there are some typos. (e.g. alliviate instead of alleviate). Could you fix any typos in the doc please?

**Comment by liferoad** [comment]
@damccorm Can you check the failed precommits? Do we need to update the author files?

**Comment by damccorm** [comment]
The errors from the website precommit are:

```
ERROR Author 'prodriguezdefino' is not defined. Please add them to 'website/www/site/data/authors.yml'
ERROR Author 'namita sharma' is not defined. Please add them to 'website/www/site/data/authors.yml'
```

so we need to add them the authors to https://github.com/apache/beam/blob/master/website/www/site/data/authors.yml as part of this PR

**Comment by damccorm** [comment]
The whitespace errors can be fixed by running `vim -es -c ":%s/\s\+$//e" -c ":wq" "./website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md"` (see the postcommit log for commands like this on failures)

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+### Source Code Notes
+
+The source code for the content discovery platform can be found available in [Github](https://github.com/prodriguezdefino/content-dicovery-platform-gcp), and can be quickly deployed in any GCP service for a test run. The repository includes source code for the integration services, the multi-language ingestion pipeline, and the deployment automation for all its components through Terraform. Consider that deployment may take up to 90 minutes to create and configure all the needed resources. Also some additional documentation regarding the deployment requisites and some example REST interactions can be found in the repository README file. 
+
+## Recap
```
Comment (rszper) [inline website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:327]:
```suggestion

```

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+This article discusses the use of Generative AI and LLM-based services to improve, automate, and add creativity to existing solutions that deal with information and data. It then describes a common use case for a content discovery platform using Google Cloud Platform services. The main goal of the use case is to exemplify how to use the available technologies and services together, serve as a quick start reference, and ease the learning curve for LLM and Generative AI capabilities, simplifying their adoption for production use cases. The code [repository](https://github.com/prodriguezdefino/content-dicovery-platform-gcp) should serve as a starting point to understand how to resolve the integration challenges and the interaction examples that build rich context for LLM requests, or resolve streaming ingestion and semantic matching, looking to really take advantage of near real time Generative AI use cases.
```
Comment (rszper) [inline website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:329]: I don't think this section is needed.

Comment (prodriguezdefino) [reply website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:332]: removed

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md (hunk 1 above)
Comment (prodriguezdefino) [reply website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md]: done

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+Also, CloudRun plays the important role of exposing the services, which will interact with the many different storage systems in use to resolve the user query requests. For example, capturing the semantic meaning from the user’s query by interacting with the embedding’s model, finding near matches from the Vertex AI Vector Search (formerly Matching Engine),which stores the embedding vectors from the ingested document’s content, retrieving the text content from BigTable to contextualize finally the request to the VertexAI Text-Bison and Chat-Bison models for a final response to the user’s originary query.
```
Comment (damccorm) [inline website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:74]: "to contextualize finally the request to the VertexAI Text-Bison and Chat-Bison models for a final response to the user’s originary query."

I don't follow what this is trying to say - is it just saying "to send the request to the VertexAI..."? I'd suggest rewording.

Also, I don't think we need to change this, but it is worth calling out that Beam has built in support for calling vertex AI endpoints (e.g. here's an example calling into a tuned vertex text-bison model - https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/inference/vertex_ai_llm_text_classification.py). Maybe we could at least call out that integration point in this blog since I imagine it would be interesting to the audience

Comment (prodriguezdefino) [reply website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:74]: Reworded the whole paragraph. 

This specific part talks about the interaction with vertex ai models from the context of the exposed services, not the Apache Beam pipeline, I will add a reference to this code example in the section with the pipeline details. Thanks!

File: website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+### Solution's Notes
+
+This solution is designed to serve as an example for learning purposes. Many of the configuration values for the extraction pipeline and security restrictions are provided only as examples. The solution doesn't propagate the existing access control lists (ACLs) of the ingested content. As a result, all users that have access to the service endpoints have access to summarizations of the ingested content from those original documents.
+
+### Notes about the source code
+
+The source code for the content discovery platform is available in [Github](https://github.com/prodriguezdefino/content-dicovery-platform-gcp). You can run it in any Google Cloud project. The repository includes the source code for the integration services, the multi-language ingestion pipeline, and the deployment automation through Terraform. If you deploy this example, it might take up to 90 minutes to create and configure all the needed resources. The README file contains additional documentation about the deployment prerequisites and example REST interactions.
```
Comment (prodriguezdefino) [inline website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:338]: Is it ok to point to personal github repositories? should this code be included in another place?

Comment (aaltay) [reply website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:338]: That is ok.  I assume that you intent to keep it there. If you prefer, you can also move it to (https://github.com/GoogleCloudPlatform/dataflow-sample-applications). If you do that, please send another PR to update this blog post.

Comment (prodriguezdefino) [reply website/www/site/content/en/blog/dyi-content-discovery-platform-genai-beam.md:338]: Yes, I will try that out on a diff PR, thanks for the clarification.

**Comment by aaltay** [comment]
LGTM. Thank you all! Merging this.

**Comment by aaltay** [comment]
Published here: https://beam.apache.org/blog/dyi-content-discovery-platform-genai-beam/
